nab-metrics/src/main/java/ru/hh/nab/metrics/CompactHistogram.java

```diff
@@ -0,0 +1,33 @@
+package ru.hh.nab.metrics;
+
+/**
+ * This class represents a histogram which compacts recorded metrics of a certain interval into a single value
+ * to avoid bloating memory with unique values by sacrificing precision.
+ */
+public class CompactHistogram extends Histogram {
+
+  /**
+   * Controls the size of a cluster and precision. Value must be power of 2.
+   */
+  private final int compactionRatio;
+
+  public CompactHistogram(int histogramSize, int compactionRatio) {
+    super(histogramSize);
+    this.compactionRatio = Integer.highestOneBit(compactionRatio);
+  }
+
+  /**
+   * Returns value after compaction procedure.
+   * <p>
+   * Recorded value 1 with compactionRatio of 30 will be compacted into cluster point 32.
+   *
+   * @param value recorded value
+   * @return compacted value
+   */
+  @Override
+  protected int calculateValue(int value) {
+    return value == 0 ? 0 : (value | compactionRatio - 1) + 1;
```

Comment:
```suggestion
    return value == 0 ? 0 : ((value - 1) | compactionRatio - 1) + 1;
```
Так не будет баги, что уже степень двойки попадает в другой разряд